lxd/db/instances.go

```diff
@@ -230,29 +220,36 @@ SELECT instances.name, nodes.id, nodes.address, nodes.heartbeat, projects.name
 var ErrInstanceListStop = fmt.Errorf("search stopped")
 
 // InstanceList loads all instances across all projects and for each instance runs the instanceFunc passing in the
-// instance and it's project and profiles. Accepts optional filter argument to specify a subset of instances.
-func (c *Cluster) InstanceList(filter *cluster.InstanceFilter, instanceFunc func(inst InstanceArgs, project api.Project) error) error {
+// instance and it's project and profiles. Accepts optional filter arguments to specify a subset of instances.
+func (c *Cluster) InstanceList(filters map[instancetype.Type][]cluster.InstanceFilter, instanceFunc func(inst InstanceArgs, project api.Project) error) error {
```

Comment:
This can just be updated to take a variadic `filters ...cluster.InstanceFilter` argument at the end.
Then inside the function if a filter.Type == instancetype.Any, set it to nil, and if the resulting filter is empty then remove it.

Reply:
I also wonder if its worth adding a function to `instancetype.Type` called `Filter() *Type` that detects if the type is `Any` and returns `nil` in that case? That would then allow directly using the instancetype to generate a filter, e.g.

```
[]cluster.InstanceFilter{{
    Type: instancetype.Any.Filter(), // Or instType.Filter() if passed in via an argument.
}}
```

Reply:
> This can just be updated to take a variadic `filters ...cluster.InstanceFilter` argument at the end. Then inside the function if a filter.Type == instancetype.Any, set it to nil, and if the resulting filter is empty then remove it.

I didn't want to do this because this will add a hardcoded nil check for each field of a struct that mostly exists for the generator inside of a non-generated function, which just adds one more thing to keep track of.

Reply:
As it stands the interface to this function seems quite odd now though. 
We can avoid the code maintainability issue though by comparing the resulting filter to an empty filter struct (default all nils), that way if new fields get added it will get automatically updated.